noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/scheduled_value_change.nr

```diff
@@ -0,0 +1,279 @@
+use dep::protocol_types::traits::{Serialize, Deserialize, FromField, ToField};
+
+// This data structure is used by SharedMutable to represent a value that changes from `pre` to `post` at some block 
+// called the `block_of_change`. The value can only be made to change by scheduling a change event at some future block
+// of change after some delay measured in blocks has elapsed. This means that at any given block number we know both the 
+// current value and the smallest block number at which the value might change - this is called the 'block horizon'.
+//
+// The delay being a type parameter instead of a struct field is an implementation detail, and is due to a number of
+// reasons:
+//  - we want to serialize and deserialize this object in order to store it in public storage, but we don't want to
+//    include the delay there because it is immutable
+//  - because of how aztec-nr state variables are declared, having a type with some immutable property is better 
+//    expressed via types, since they are always constructed with the same `::new(context, storage_slot)` function.
+struct ScheduledValueChange<T, DELAY> {
+    pre: T,
+    post: T,
+    block_of_change: u32,
+    // The _dummy variable forces DELAY to be interpreted as a numberic value. This is a workaround to
+    // https://github.com/noir-lang/noir/issues/4633. Remove once resolved.
+    _dummy: [Field; DELAY],
+}
+
+impl<T, DELAY> ScheduledValueChange<T, DELAY> {
+    pub fn new(pre: T, post: T, block_of_change: u32) -> Self {
+        Self { pre, post, block_of_change, _dummy: [0; DELAY] }
+    }
+
+    pub fn get_current_at(self, historical_block_number: u32) -> T {
```

Comment:
Something feels weird here, that if I read really far back, the "pre" might not be the actual value. 

Like. At block 10 value might be `(a,b,c)` and then later it becomes `(b,d,e)` but If I then say read at 10, I would read the pre as `b`. 

It makes sense, was just from the naming that it is what sprung to mind.

Reply:
I'm not sure I follow your example, could you provide concrete values for `c` and `e`?

But yes, values do move from `post` to `pre` when a new change is scheduled. Clients that have not seen the new scheduled change will read `post`, while clients that have seen the change will read `pre`. The important thing is that these values will be the same, and all reads will be valid (as long as `max_block_number` is properly constrained).

This is what the `assert_block_horizon_invariants` helper function tests: `current_value` returns the same value at the historical block number and at the block horizon, even if a new change is scheduled as early as possible (i.e. in the block following the historical block).